mlrun/model_monitoring/db/tsdb/timescaledb/timescaledb_connection.py

```diff
@@ -0,0 +1,424 @@
+        """Execute a query and return formatted results."""
+        if isinstance(query, Statement):
+            query.execute(cursor)
+        else:
+            cursor.execute(query)
+
+        if cursor.description:
+            field_names = [desc.name for desc in cursor.description]
+            results = cursor.fetchall()
+            data = [tuple(row) for row in results]
+            return QueryResult(data, field_names)
+        else:
+            return QueryResult([], [])
+
+    def execute_with_fallback(
+        self,
+        pre_aggregate_manager: PreAggregateManager,
+        pre_agg_query_builder: Callable[[], str],
+        raw_query_builder: Callable[[], str],
+        interval: Optional[str] = None,
+        agg_funcs: Optional[list[str]] = None,
+        column_mapping_rules: Optional[dict[str, list[str]]] = None,
+        debug_name: str = "query",
+    ) -> pd.DataFrame:
+        """
+        Execute a query with pre-aggregate optimization and automatic fallback.
+
+        This method encapsulates the common pattern of trying pre-aggregate queries first,
+        then falling back to raw data queries if the pre-aggregate fails.
+
+        :param pre_aggregate_manager: Manager for pre-aggregate operations
+        :param pre_agg_query_builder: Function that returns pre-aggregate query string
+        :param raw_query_builder: Function that returns raw data query string
+        :param interval: Time interval for aggregation
+        :param agg_funcs: List of aggregation functions
+        :param column_mapping_rules: Rules for mapping column names in pre-aggregate results
+        :param debug_name: Name for debugging/logging purposes
+        :return: DataFrame with query results
+        """
+        # Import locally to avoid circular dependency
+        from mlrun.model_monitoring.db.tsdb.timescaledb.utils.timescaledb_dataframe_processor import (
+            TimescaleDBDataFrameProcessor,
+        )
+
+        df_processor = TimescaleDBDataFrameProcessor()
+
+        if pre_aggregate_manager.can_use_pre_aggregates(
+            interval=interval, agg_funcs=agg_funcs
+        ):
+            try:
+                # Try pre-aggregate query first
+                query = pre_agg_query_builder()
+                result = self.run(query=query)
+                df = df_processor.from_query_result(result)
+
+                if not df.empty and column_mapping_rules:
+                    # Apply flexible column mapping for pre-aggregate results
+                    mapping = df_processor.build_flexible_column_mapping(
+                        df, column_mapping_rules
+                    )
+                    df = df_processor.apply_column_mapping(df, mapping)
+
+                return df
+
+            except Exception as e:
+                logger.warning(
+                    f"Pre-aggregate {debug_name} query failed, falling back to raw data: {e}",
```

Comment:
We don't want to log the error twice.
```suggestion
                    f"Pre-aggregate {debug_name} query failed, falling back to raw data",
```

Reply:
will fix